### PR TITLE
Implement ToDoService

### PR DESCRIPTION
**Pull request by hieuvuong2310**

### CODE REVIEW

File: ToDoApp/ToDoApp/ContentView.swift [hunk 1]
```diff
@@ -6,7 +6,6 @@
 //
 
 import SwiftUI
-
```
Comment (vladyslavsosiuk) [inline ToDoApp/ToDoApp/ContentView.swift:9]: it is good practice to not make changes which are not related to a PR. for example, this new line. since a bunch of such changes may keep a reviewer unfocused

File: ToDoApp/ToDoApp/ToDoService.swift [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+enum TaskError: Error {
+    case unavailableTask
+}
+
+struct Task {
+    var id: String = ""
+    var name: String = ""
+    var deadline: Date = Date.now
+    var status: Bool = false
```
Comment (vladyslavsosiuk) [inline ToDoApp/ToDoApp/ToDoService.swift]: do we really wand id to be a variable? should we ever change it?
do we really want to create a Task with a default id value of empty string? I suggest to use UUID type to represent Id of a task
do we really want to create a Task without a name with default value of empty string? 
since deadline could be optional shouldn't it's default value be nil?

File: ToDoApp/ToDoApp/ToDoService.swift [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    init(name: String, deadline: Date?) {
```
Comment (vladyslavsosiuk) [inline ToDoApp/ToDoApp/ToDoService.swift]: suggestion: Swift automatically synthesizes initializers for structs, so in most cases you don't need to write it yourself

File: ToDoApp/ToDoApp/ToDoService.swift [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        self.name = name
+        self.deadline = deadline!
```
Comment (vladyslavsosiuk) [inline ToDoApp/ToDoApp/ToDoService.swift]: there should be a good reason to use force unwrap, currently I don't see it, since nil may be passed to the initializer and the app will crash

File: ToDoApp/ToDoApp/ToDoService.swift [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    }
+}
+
+protocol ToDoService {
+    func createTask(name: String, deadline: Date?) -> Task
+    func updateTask(todo: Task) throws
+    func taskForToday() -> [Task]
```
Comment (vladyslavsosiuk) [inline ToDoApp/ToDoApp/ToDoService.swift]: maybe you would like to introduce async interfaces here, since when you will connect a cloud service you won't refactor client code which already use ToDoService

you can choose either old fashion closure interfaces or new async interfaces
`func createTask(name: String, deadline: Date?, completion: @escaping (Result<Task, TaskError>) -> Void)`
or
`func createTask(name: String, deadline: Date?) async -> Result<Task, TaskError>`

File: ToDoApp/ToDoApp/ToDoService.swift [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        return oneTask
+    }
+    // Update the task
+    func updateTask(todo: Task) throws {
+        guard let index = tasks.firstIndex(where: {
+            $0.id == todo.id
+        })else {
+            throw TaskError.unavailableTask
+        }
+        tasks.insert(todo, at: index)
```
Comment (vladyslavsosiuk) [inline ToDoApp/ToDoApp/ToDoService.swift]: should be tasks[index] = todo

File: ToDoApp/ToDoApp/ToDoService.swift [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    }
+    // Get list of tasks for today
+    func taskForToday() -> [Task] {
+        let date = Date.now
+        let todayList = tasks.filter({$0.deadline == date})
```
Comment (vladyslavsosiuk) [inline ToDoApp/ToDoApp/ToDoService.swift]: [Swift Date type](https://developer.apple.com/documentation/foundation/date) represent a time interval relative to an absolute reference date in seconds. This means that when Date.now is returning its value it returned time interval in seconds from 1970 to the time it has been invoked. That's why comparing deadline date with now will not work. in order to check if date's day is equal to today you need to use Calendar and its method isDateInToday https://developer.apple.com/documentation/foundation/calendar/2293243-isdateintoday I suggest to create a protocol which will have the same signature for the isDayInTodayMethod, conform Calendar to this protocol and inject it as a dependency so you could test the ToDoService implementation later

**Review by vladyslavsosiuk — CHANGES_REQUESTED** [review]

File: ToDoApp/ToDoApp/ToDoService.swift (hunk 2 above)
Comment (hieuvuong2310) [reply ToDoApp/ToDoApp/ToDoService.swift]: Oh I thought if a task without any deadline, it'll be set for today as the default deadline. I will change it to nil now.
And for a task without a name, I think I'll put the default value named "New Task." I'm fixing it now.

File: ToDoApp/ToDoApp/ToDoService.swift (hunk 4 above)
Comment (hieuvuong2310) [reply ToDoApp/ToDoApp/ToDoService.swift]: This has been fixed.

File: ToDoApp/ToDoApp/ToDoService.swift (hunk 6 above)
Comment (hieuvuong2310) [reply ToDoApp/ToDoApp/ToDoService.swift]: This has been fixed.

File: ToDoApp/ToDoApp/ToDoService.swift (hunk 5 above)
Comment (hieuvuong2310) [reply ToDoApp/ToDoApp/ToDoService.swift]: This has been fixed.

File: ToDoApp/ToDoApp/ToDoService.swift (hunk 2 above)
Comment (vladyslavsosiuk) [reply ToDoApp/ToDoApp/ToDoService.swift]: But why do we need a default string for the name? A task will be always created with a name provided by the user
I also believe that the note should be either required to be provided by the user or be optional since I don't see a case when the user would like to always provide a note. for example what would be the note if the task is just to by a coffee

here is an example: the id is let, since it will never change, for title there is no default value, since the user will always provide own value, note is optional since it is not required.
```
struct Task {
     let id: UUID = UUID()
     var title: String
     var note: String?
     var deadline: Date?
     var status: Bool = false
 }
```

The last thing is the instantiation of the id. It would be better to carry over the responsibility of creating the id to another object, to be able to properly test task creation, but I think we can skip this for now

File: ToDoApp/ToDoApp/ToDoService.swift [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    var title: String = "New Task"
+    var note: String = "New Note"
+    var deadline: Date?
+    var status: Bool = false
+}
+
+protocol ToDoService {
+    func createTask(title: String, note: String, deadline: Date?) async -> Result<Task, TaskError>
+    func updateTask(todo: Task) async -> Result<Task, TaskError>
+    func taskForToday() -> [Task]
```
Comment (vladyslavsosiuk) [inline ToDoApp/ToDoApp/ToDoService.swift]: this method should be async as well since tasks will be fetched

File: ToDoApp/ToDoApp/ToDoService.swift [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    func checkIsToday(date: Date?) -> Bool
```
Comment (vladyslavsosiuk) [inline ToDoApp/ToDoApp/ToDoService.swift]: this method is not needed to be in the ToDoService protocol, since clients of the ToDoService don't need to invoke this method, and it should be private in the FeaturesToDo class

File: ToDoApp/ToDoApp/ToDoService.swift [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        guard let index = tasks.firstIndex(where: {
+            $0.id == todo.id
+        })else {
+            return Result.failure(TaskError.unavailableTask)
+        }
+        tasks[index] = todo
+        return Result.success(todo)
+    }
+    // Get list of tasks for today
+    func taskForToday() -> [Task] {
```
Comment (vladyslavsosiuk) [inline ToDoApp/ToDoApp/ToDoService.swift]: there a couple of warnings in those methods

**Review by vladyslavsosiuk — CHANGES_REQUESTED** [review]

File: ToDoApp/ToDoApp/ToDoService.swift [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        let date = Date.now
+        let todayList = tasks.filter({checkIsToday(date: $0.deadline ?? nil)})
+        return todayList
+    }
+    // Check whether the deadline is today
+    func checkIsToday(date: Date?) -> Bool {
+        var cal: Calendar = Calendar(identifier: .hebrew)
```
Comment (vladyslavsosiuk) [inline ToDoApp/ToDoApp/ToDoService.swift]: let's use Calendar.current instead of the Hebrew calendar

**Review by vladyslavsosiuk — CHANGES_REQUESTED** [review]

File: ToDoApp/ToDoApp/ToDoService.swift [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+//
+
+import Foundation
+
+enum TaskError: Error {
+    case unavailableTask
+}
+
+struct Task {
+    var id: UUID = UUID()
```
Comment (vladyslavsosiuk) [inline ToDoApp/ToDoApp/ToDoService.swift]: id should be a constant since we will never change it

File: ToDoApp/ToDoApp/ToDoService.swift [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+protocol DateChecker {
+    func isDateInToday(_ date: Date) -> Bool
+}
+//  conform Calendar to this protocol
+extension Calendar: DateChecker {}
+
+protocol ToDoService {
+    func createTask(title: String, note: String, deadline: Date?) async -> Result<Task, TaskError>
+    func updateTask(todo: Task) async -> Result<Task, TaskError>
+    func taskForToday() async -> [Task]
```
Comment (vladyslavsosiuk) [inline ToDoApp/ToDoApp/ToDoService.swift]: let's have this returning a Result as it may also fail

**Review by vladyslavsosiuk — CHANGES_REQUESTED** [review]

File: ToDoApp/ToDoApp/ToDoService.swift (hunk 13 above)
Comment (hieuvuong2310) [reply ToDoApp/ToDoApp/ToDoService.swift]: This has been fixed.

File: ToDoApp/ToDoApp/ToDoService.swift (hunk 12 above)
Comment (hieuvuong2310) [reply ToDoApp/ToDoApp/ToDoService.swift]: This has been fixed.

**Review by vladyslavsosiuk — APPROVED** [review]
